File: src/main/java/maestro/behavior/Behavior.java
package maestro.behavior;

import maestro.Maestro;
import maestro.api.behavior.IBehavior;
import maestro.api.utils.IPlayerContext;

/**
 * A type of game event listener that is given {@link Maestro} instance context.
 *
 * @author Brady
 * @since 8/1/2018
 */
public class Behavior implements IBehavior {

    public final Maestro maestro;
    public final IPlayerContext ctx;

    protected Behavior(Maestro maestro) {
        this.maestro = maestro;
        this.ctx = maestro.getPlayerContext();
    }
}

File: src/main/java/maestro/behavior/Behavior.kt
package maestro.behavior

import maestro.Agent
import maestro.api.behavior.IBehavior
import maestro.api.utils.IPlayerContext

/**
 * Base behavior class that provides [Agent] and [IPlayerContext] to subclasses.
 *
 * All behaviors have access to the agent instance and player context through protected fields.
 *
 * @author Brady
 * @since 8/1/2018
 */
abstract class Behavior protected constructor(
    @JvmField val maestro: Agent
) : IBehavior {
    @JvmField
    val ctx: IPlayerContext = maestro.getPlayerContext()
}
